### ISSUE TITLE
Race condition likely during unsubscribe

### ISSUE THREAD

**Issue by rudyjantz** [nats-io/cnats#184]
Hi @kozlovic ,

You handled a previous (invalid) test case from me:
https://github.com/nats-io/cnats/issues/154

After continuing to experience problems, though, I updated and improved the tests, and I think I found that the original issue may indeed exist.

Attached are the updated test case. To use:
```
# ensure gnatsd is running, and then ...
make
./test_cnats.out 0
./test_cnats.out 1
./test_cnats.out 2
# etc...
```

What I've found is that tests 9 and 10 fail. A workaround is to introduce a sleep between `natsSubscription_Unsubscribe` and `natsSubscription_Destroy` (line 99 in the provided cnats_iface.c).

I haven't identified the fix yet. It seems the gLib ref count doesn't get decrement completely, perhaps because the thread in natsSub_deliverMsg doesn't always break out at
```
if (sub->closed)
{
    natsSub_Unlock(sub);
    break;
}
```

I'll investigate further if time permits.

Thanks.

[cnats_connect_disconnect_issue.tar.gz](https://github.com/nats-io/cnats/files/2453355/cnats_connect_disconnect_issue.tar.gz)

**Comment by rudyjantz** [comment]
This is with cnats 1.7.6.

**Comment by rudyjantz** [comment]
Right, the previous tests can be ignored. For these tests, it's your call.

The issue is this: An application written atop cnats must call `nats_Close()` in order to clean up all of its memory, but reconnecting after a call to `nats_Close()` may fail.

So as a developer of a cnats application, I have two options:
1. Don't call `nats_Close()` ever, in which case the memory footprint at process exit is ~35KB bigger than it should be (on this test machine). That could impact a production system, but it's also a development problem in terms of verifying other memory leaks.
2. Call `nats_Close()` only on application exit. This is unrealistic for my scenarios.

Regarding the documentation and usage of `nats_Open()` and `nats_Close()`: Right, I've looked through this, and I've read the point about not using them concurrently, which is well taken. In this test case, however, it's arguably _not_ concurrent, though -- not from the standpoint of the developer, that is. The test is single-threaded. That it's multithreaded under the hood is not known to the developer. And if it's required for the developer to know that, then as a developer I need a hook from cnats in order to know when `nats_Close()` has completed.

I get your point about these test cases exercising the open/close functions not as they "were designed for". This would seem a pathological case. But you may notice that test-9 can fail within just 1 or 2 iterations. And that's exactly what can happen in my application. My application may "think" it's done with cnats and choose to tear down, but then it may immediately require a new connection. That's reasonable, and test-9 distills it. Ref-counting cnats connections within my application is not enough to fix this. I either have to add an unwanted 500us sleep into my disconnect; I have to patch cnats; or I have to allow ~35KB per cnats process on my system, even when no connections are open.

**Comment by kozlovic** [comment]
I could introduce a new close API that blocks waiting for all threads holding a reference to the library to return. This would obviously prevent the call on that new API from within a NATS thread (user's message callback, async callbacks, etc..).

But what I find a bit concerning of the usage that you describe is how inefficient that could be. Tearing down the library will stop supporting threads that will need to be recreating at the next connect.
I don't think it is unreasonable to do an open/close at beginning and end of application. The library would be loaded once, and cleaned-up once. Note that a connect will implicitly open the library if not already done, but library does not have to be closed after a connection close/destroy. Again, creating multiple connections does not load (open) the library multiple times.

I am curious as to the reason for your statement: `Call nats_Close() only on application exit. This is unrealistic for my scenarios.`. Could you elaborate?

**Comment by kozlovic** [comment]
@rudyjantz I have PR #187 that would introduce a new API, nats_CloseAndWait() that will ensure that all NATS threads and resources are released before returning. Since this call would block, it can't be invoked from any NATS thread, which NATS callbacks are invoked from. If that happens, an error is returned and the library is not closed. Is that something that would work for you?

**Comment by rudyjantz** [comment]
@kozlovic Nice hacking. I was sitting on a response while considering what else I could do from my end. But this solves the problem for me. I tested PR 187 with test-9 from the above tarball, and it runs smoothly.

A few comments regarding all this:
- You had to add a new API call, which exposes you to potentially more bugs and maintenance, and even misuse from folks who will take the performance penalty unknowingly. These are all valid reasons to hesitate before making the change, which is why I was thinking through things on my end. But you ended up getting it out, so thank you.
- `Calling nats_Close() only on application exit. This is unrealistic for my scenarios.` This is a little complicated. The short of it is that I've wrapped cnats, and clients don't have to worry about closing. But for that to happen, the wrapper has to close cnats when the last connection is closed. The PR you just wrote addresses the case when the user quickly tries to reconnect.
- In the long run, I should able to move off of nats_CloseAndWait() and on to nats_Close(). Clients shouldn't be constantly opening/closing connections, as you pointed out, because they're taking an unnecessary performance hit. But I don't have total control over that, so this gives me a stable option in the interim.

I have a separate issue that I'll open regarding publishing. It's related, and I thought this PR might solve it.

Thanks.

**Comment by kozlovic** [comment]
@rudyjantz See my comments:

> You had to add a new API call,

If we want to have the call block, I had to introduce a new API otherwise that may possibly break users calling `nats_Close()` from a NATS thread. What I could do, though, is not introduce the new one and since I can detect if `nats_Close()` is called from a NATS thread or not, is to behave as it was before if called from a NATS thread (that is, still tear down the library but not wait), or wait if it knows that it is not called from such thread. That would prevent adding a new API.

> But for that to happen, the wrapper has to close cnats when the last connection is closed. The PR you just wrote addresses the case when the user quickly tries to reconnect.

Understood.

> In the long run, I should be able to move off of nats_CloseAndWait() and on to nats_Close(). Clients shouldn't be constantly opening/closing connections, as you pointed out, because they're taking an unnecessary performance hit. But I don't have total control over that, so this gives me a stable option in the interim.

Glad that you understand the impact of opening/closing the library too often.

I really lean toward having `nats_Close()` updated to behave as I described above: block/wait if it can (not invoked from NATS thread), not wait but still tear down the library if invoked from library - as it does today.

> There's another related "issue" to all this, and I was hoping your PR would fix it. Unfortunately it doesn't, so I'll open a separate github issue on it. (Update: see issue #188)

Will comment there.

**Comment by rudyjantz** [comment]
Right, updating `nats_Close()` (rather than adding a new call) might be a good option. It'll be backwards compatible. It sounds like it'll behave reasonably in all scenarios.

**Comment by kozlovic** [comment]
Actually, after putting more thoughts into it, it is safer to have a new call than change behavior of `nats_Close()`.

The problem is that if I change `nats_Close()` to now wait (assuming not invoked from a NATS thread), but existing users had applications that were calling `nats_Close()` and later destroyed some of the NATS objects (which arguably they should do *before*), then this call would block for ever.
In other words, say that you create a connection, call `nats_Close()` and then only after that call `natsConnection_Destroy()`, then this code would work in master but "deadlock" with the change.

Introducing the new API call (and documenting this behavior) would avoid possibly breaking existing applications.

**Comment by kozlovic** [comment]
@rudyjantz I have updated (push force) the PR with adding a timeout to the new `nats_CloseAndWait()` API, and added/enhanced the doc for the two close APIs.

**Comment by rudyjantz** [comment]
Very nice. I tested with timeout -1, and it went well. With timeout == 0, I had issues again on test 9 that I sent. The docs state
```
 * @param timeout the maximum time to wait for the library to be closed. If
 * negative or 0, waits for as long as needed.```

If I'm reading that right, then this line should be <= 0?
https://github.com/nats-io/cnats/pull/187/commits/1e48414fdb75687f98c73341117f603d3c00b01b#diff-a4429a58019ab474894d90837d1a6051R1195

**Comment by kozlovic** [comment]
Good catch! I will fix that.
